src/domain/account.ts

```diff
@@ -1,264 +1,318 @@
+    if (!this.walletRepository) throw new Error('No wallet repository, cannot get next address');
+    if (!this.name) throw new Error('No name, cannot get next address');
 
-export const initialCustomRestorerOpts: CustomRestorerOpts = {
-  lastUsedExternalIndex: -1,
-  lastUsedInternalIndex: -1,
-  customParamsByIndex: {},
-  customParamsByChangeIndex: {},
-};
+    const lastUsedIndexes = await this.getLastUsedIndexes();
+    const lastUsed = isInternal ? lastUsedIndexes.internal : lastUsedIndexes.external;
+    const scripts = await this.deriveBatch(lastUsed, lastUsed + 1, isInternal);
+    const script = scripts[0];
+    const { publicKey } = this.deriveBlindingKey(script);
+    const address = payments.p2wpkh({
+      output: script,
+      network: this.network,
+      blindkey: publicKey,
+    }).confidentialAddress;
+    if (!address) throw new Error('Could not derive address');
+    // increment the account details
+    await this.walletRepository.updateAccountLastUsedIndexes(
+      this.name,
+      this.network.name as NetworkString,
+      { [isInternal ? 'internal' : 'external']: lastUsed + 1 }
+    );
 
-function makeRestorerFromChainAPI<T extends IdentityInterface>(
-  id: T,
-  getAddress: (isChange: boolean, index: number) => string
-): Restorer<{ api: ChainAPI; gapLimit: number }, IdentityInterface> {
-  return async ({ gapLimit, api }) => {
-    const restoreFunc = async function (
-      getAddrFunc: (index: number) => Promise<string>
-    ): Promise<number | undefined> {
-      let counter = 0;
-      let next = 0;
-      let maxIndex: number | undefined = undefined;
-
-      while (counter < gapLimit) {
-        const cpyNext = next;
-        // generate a set of addresses from next to (next + gapLimit - 1)
-        const addrs = await Promise.all(
-          Array.from(Array(gapLimit).keys())
-            .map((i) => i + cpyNext)
-            .map(getAddrFunc)
-        );
+    return address;
+  }
 
-        const hasBeenUsedArray = await api.addressesHasBeenUsed(addrs);
+  async sync(gapLimit = GAP_LIMIT): Promise<{
+    lastUsed: { internal: number; external: number };
+  }> {
+    if (!this.chainSource) throw new Error('No chain source, cannot sync');
+    if (!this.walletRepository) throw new Error('No wallet repository, cannot sync');
+    if (!this.name) throw new Error('No name, cannot sync');
 
-        let indexInArray = 0;
-        for (const hasBeenUsed of hasBeenUsedArray) {
-          if (hasBeenUsed) {
-            maxIndex = indexInArray + next;
-            counter = 0;
+    const historyTxsId: Set<string> = new Set();
+    const txidHeight: Map<string, number | undefined> = new Map();
+
+    const lastUsed = await this.getLastUsedIndexes();
+
+    const walletChains = [0, 1];
+    for (const i of walletChains) {
+      const isInternal = i === 1;
+      let batchCount = isInternal ? lastUsed.internal : lastUsed.external;
+      let unusedScriptCounter = 0;
+
+      while (unusedScriptCounter <= gapLimit) {
```

Comment:
I think there is a bug here. We loop one extra time when having no histories.
Say lastUsedExternalIndex is 3 and gapLimit is 3, we derive scripts/histories for 3 to 6, then  unusedScriptCounter is 3, unusedScriptCounter <= gapLimit is true, we derive again for 6 to 9.

Should it be `unusedScriptCounter < gapLimit` instead?
@louisinger

Reply:
Both of them produce the same restore result on my wallet. I would keep the "=" (better to test 1 additional batch instead of missing some txs. Note that if the next tested batch does not have histories, the scripts won't be restored).